PixiEditor/Views/UserControls/Zoombox.xaml.cs

```diff
@@ -0,0 +1,274 @@
+            sender.activeDragOperation?.Terminate();
+            sender.activeDragOperation = null;
+            sender.activeMouseDownEventArgs = null;
+        }
+
+        public Zoombox()
+        {
+            InitializeComponent();
+        }
+
+        public void CenterContent()
+        {
+            const double marginFactor = 1.1;
+            double scaleFactor = Math.Max(
+                mainGrid.ActualWidth * marginFactor / mainCanvas.ActualWidth,
+                mainGrid.ActualHeight * marginFactor / mainCanvas.ActualHeight);
+            ZoomPower = -Math.Log(scaleFactor, zoomFactor);
+            SpaceOriginPos = new Point(
+                mainCanvas.ActualWidth / 2 - mainGrid.ActualWidth * Zoom / 2,
+                mainCanvas.ActualHeight / 2 - mainGrid.ActualHeight * Zoom / 2);
+        }
+
+        public void ZoomIntoCenter(double delta)
+        {
+            ZoomInto(new Point(mainCanvas.ActualWidth / 2, mainCanvas.ActualHeight / 2), delta);
+        }
+        public void ZoomInto(Point mousePos, double delta)
+        {
+            var oldZoomboxMousePos = ToZoomboxSpace(mousePos);
+
+            ZoomPower += delta;
+
+            if (Math.Abs(ZoomPower) < 1) ZoomPower = 0;
+
+            var shiftedMousePos = ToScreenSpace(oldZoomboxMousePos);
+            var deltaMousePos = mousePos - shiftedMousePos;
+            SpaceOriginPos = SpaceOriginPos + deltaMousePos;
+        }
+
+        private Point ToScreenSpace(Point p)
+        {
+            double zoom = Zoom;
+            p.X *= zoom;
+            p.Y *= zoom;
+            p += (Vector)SpaceOriginPos;
+            return p;
+        }
+
+        private Point ToZoomboxSpace(Point mousePos)
+        {
+            double zoom = Zoom;
+            mousePos -= (Vector)SpaceOriginPos;
+            mousePos.X /= zoom;
+            mousePos.Y /= zoom;
+            return mousePos;
+        }
+
+        private void OnMouseDown(object sender, MouseButtonEventArgs e)
+        {
+            if (e.ChangedButton == MouseButton.Right)
+                return;
+            activeMouseDownEventArgs = e;
+            activeMouseDownPos = e.GetPosition(mainCanvas);
+            Keyboard.Focus(this);
+        }
+
+        private void InitiateDrag(MouseButtonEventArgs e)
+        {
+            if (ZoomMode == Mode.Normal)
+                return;
+
+            activeDragOperation?.Terminate();
+
+            if (ZoomMode == Mode.MoveTool)
+                activeDragOperation = new MoveDragOperation(this);
+            else if (ZoomMode == Mode.ZoomTool)
+                activeDragOperation = new ZoomDragOperation(this);
+
+            activeDragOperation.Start(e);
+        }
+
+        private void OnMouseUp(object sender, MouseButtonEventArgs e)
+        {
+            if (e.ChangedButton == MouseButton.Right)
+                return;
+            if (activeDragOperation != null)
+            {
+                activeDragOperation?.Terminate();
```

Comment:
No null check is required here as it's already done 2 lines above

Reply:
yep